loki/transformations/single_column/vertical.py

```diff
@@ -0,0 +1,218 @@
+        argument_names = [arg.name.lower() for arg in routine.arguments]
+        local_arrays = [arr for arr in arrays if arr.name.lower() not in argument_names]
+        # only those with the vertical size within shape
+        relevant_local_arrays = [arr for arr in local_arrays if self.vertical.size.lower()
+                in [var.name.lower() for var in FindVariables().visit(arr.shape)]]
+        # filter arrays to be ignored (for whatever reason)
+        ignore_names = self.find_local_arrays_to_be_ignored(routine)
+        if ignore_names:
+            relevant_local_arrays = [arr for arr in relevant_local_arrays if arr.name.lower() not in ignore_names]
+        return relevant_local_arrays
+
+    def find_local_arrays_to_be_ignored(self, routine):
+        """
+        Identify variables to be ignore regarding demotion for whatever reason.
+
+        Reasons are:
+
+        * explicitly marked to be ignored via pragmas within the input source file, e.g.,
+          'loki k-caching ignore(var1, var2, ...)'
+        """
+        ignore = ()
+        pragmas = FindNodes(ir.Pragma).visit(routine.body)
+        # look for 'loki k-caching ignore(var1, var2, ...)' pragmas within routine and ignore those vars
+        for pragma in pragmas:
+            if is_loki_pragma(pragma, starts_with='k-caching'):
+                pragma_ignore = get_pragma_parameters(pragma, starts_with='k-caching').get('ignore', None)
+                if pragma_ignore:
+                    ignore += as_tuple(v.strip() for v in pragma_ignore.split(','))
+        ignore_names = set(var.lower() for var in ignore)
+        return ignore_names
+
+    def identify_multilevel_arrays(self, local_arrays):
+        """
+        Identify local arrays/temporaries that have an access in the vertical dimension
+        that is different to '<vertical.index>', e.g., '<vertical.index> +/- 1'
+        """
+        multilevel_local_arrays = []
+        for arr in local_arrays:
+            for dim in arr.dimensions:
+                if self.vertical.index in FindVariables().visit(dim):
+                    # dim is not equal to vertical.index e.g., vertical.index +/- 1
+                    if dim != self.vertical.index:
+                        multilevel_local_arrays.append(arr)
+        return multilevel_local_arrays
+
+    def correct_init_of_multilevel_arrays(self, routine, multilevel_local_arrays):
+        """
+        Possibly handle initaliztion of those multilevel local arrays via
+        splitting relevant loops or rather creating a new node with the relevant
+        nodes moved to the newly created loop.
+
+        .. note::
+            This relies on pragmas being inserted in the input source code!
+        """
+        loop_map = {}
+        # find/identify loops with pragma 'loop-fusion group(<group-name>-init)'
+        with pragmas_attached(routine, ir.Loop):
```

Comment:
Should we make `pragmas_attached` a prerequisite for all the utilities and do it once, in `process_kernel`, maybe? If so, this should be mentioned in the utility method docstrings though.

Reply:
Yeah, that's true too. As the `Transformer` object behind them are reasonably fast, I'm happy to keep this as is (and as you say, it probably won't stay like this for long).